File: back-end/src/main/java/pl/straburzynski/campfiresongs/session/service/SessionService.java
package pl.straburzynski.campfiresongs.session.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import pl.straburzynski.campfiresongs.session.exception.NotAuthorizedException;
import pl.straburzynski.campfiresongs.session.model.Session;
import pl.straburzynski.campfiresongs.session.model.SessionDto;
import pl.straburzynski.campfiresongs.session.repository.SessionRepository;
import pl.straburzynski.campfiresongs.session.exception.SessionNotFoundException;
import pl.straburzynski.campfiresongs.song.service.SongConverter;

import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

@Service
@Slf4j
@RequiredArgsConstructor
public class SessionService {

    private final SessionRepository sessionRepository;
    private final SimpMessagingTemplate template;
    private final SongConverter songConverter;
    private final SessionConverter sessionConverter;
    private final PasswordEncoder passwordEncoder;

    public SessionDto findBySessionName(String sessionName) {
        Session session = sessionRepository.findByName(sessionName)
                .orElseThrow(() -> new SessionNotFoundException(sessionName));
        return sessionConverter.convertFromSession(session);
    }

    public SessionDto createSession(SessionDto sessionDto) {
        Session session;
        Optional<Session> foundSession = sessionRepository.findByName(sessionDto.getName());
        if (foundSession.isPresent()) {
            checkPassword(sessionDto, foundSession.get());
            session = foundSession.get();
        } else {
            session = sessionRepository.save(
                    Session.builder()
                            .id(UUID.randomUUID())
                            .name(sessionDto.getName())
                            .password(passwordEncoder.encode(sessionDto.getPassword()))
                            .build()
            );
        }
        log.debug("Create session, id: {}, name: {}", session.getId(), session.getName());
        return sessionConverter.convertFromSession(session);
    }

    private void checkPassword(SessionDto sessionDto, Session session) {
        if (Objects.isNull(sessionDto.getPassword()) || !passwordEncoder.matches(sessionDto.getPassword(), session.getPassword())) {
            throw new NotAuthorizedException();
        }
    }

    public SessionDto updateSession(SessionDto sessionDto) {
        Session foundSession = sessionRepository.findByName(sessionDto.getName())
                .orElseThrow(() -> new SessionNotFoundException(sessionDto.getName()));
        foundSession.setTemporary(sessionDto.isTemporary());
        if (sessionDto.isTemporary()) {
            foundSession.setSong(null);
            foundSession.setSongData(songConverter.stringifyFromSongDto(sessionDto.getSong()));
        } else {
            foundSession.setSong(songConverter.convertFromSongDto(sessionDto.getSong()));
            foundSession.setSongData(null);
        }
        Session updatedSession = sessionRepository.save(foundSession);
        log.debug("Update session, name: {}, temporary: {}, songId: {}",
                updatedSession.getName(), updatedSession.isTemporary(),
                updatedSession.isTemporary() ? "---" : updatedSession.getSong().getId());

        template.convertAndSend("/topic/message/" + updatedSession.getName(), sessionDto.getSong());
        return sessionConverter.convertFromSession(updatedSession);
    }

    public void deleteSession(UUID id) {
        Session foundSession = sessionRepository.findById(id).orElseThrow(() -> new SessionNotFoundException(id.toString()));
        log.debug("Deleting session, id: {}, name: {}", foundSession.getId(), foundSession.getName());
        sessionRepository.deleteById(id);
    }

}

File: back-end-kotlin/src/main/kotlin/pl/straburzynski/campfiresongs/session/service/SessionService.kt
package pl.straburzynski.campfiresongs.session.service

import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import pl.straburzynski.campfiresongs.session.exception.NotAuthorizedException
import pl.straburzynski.campfiresongs.session.exception.SessionNotFoundException
import pl.straburzynski.campfiresongs.session.model.Session
import pl.straburzynski.campfiresongs.session.model.SessionDto
import pl.straburzynski.campfiresongs.session.repository.SessionRepository
import pl.straburzynski.campfiresongs.song.service.SongConverter
import java.util.UUID
import java.util.logging.Logger
import kotlin.jvm.optionals.getOrNull


@Service
class SessionService(
    val sessionRepository: SessionRepository,
    val template: SimpMessagingTemplate,
    val songConverter: SongConverter,
    val sessionConverter: SessionConverter,
    val passwordEncoder: PasswordEncoder
) {
    fun findBySessionName(sessionName: String): SessionDto {
        val session: Session = sessionRepository.findByName(sessionName) ?: throw SessionNotFoundException(sessionName)
        return sessionConverter.convertFromSession(session)
    }

    fun createSession(sessionDto: SessionDto): SessionDto {
        val foundSession = sessionRepository.findByName(sessionDto.name)
        val session = if (foundSession != null) {
            checkPassword(sessionDto, foundSession)
            foundSession
        } else sessionRepository.save(
            Session(
                name = sessionDto.name,
                password = passwordEncoder.encode(sessionDto.password)
            )
        )
        logger.info("Create session, id: ${session.id}, name: ${session.name}")
        return sessionConverter.convertFromSession(session)
    }

    private fun checkPassword(sessionDto: SessionDto, session: Session) {
        if (sessionDto.password == null ||
            !passwordEncoder.matches(sessionDto.password, session.password)
        ) throw NotAuthorizedException()
    }

    fun updateSession(sessionDto: SessionDto): SessionDto {
        val sessionId = sessionDto.id ?: throw SessionNotFoundException()
        val foundSession = sessionRepository.findById(sessionId).getOrNull()
            ?: throw SessionNotFoundException(sessionDto.name)
        foundSession.temporary = sessionDto.temporary
        if (sessionDto.temporary) {
            foundSession.song = null
            foundSession.songData = sessionDto.song?.let { songConverter.stringifyFromSongDto(sessionDto.song) }
        } else {
            foundSession.song = sessionDto.song?.let { songConverter.convertFromSongDto(sessionDto.song) }
            foundSession.songData = null
        }
        val updatedSession = sessionRepository.save(foundSession)
        logger.info(
            "Update session, name: ${updatedSession.name}, temporary: ${updatedSession.temporary}, " +
                    "songId: ${if (updatedSession.temporary) "---" else updatedSession.song?.id}"
        )
        updatedSession.song?.let {
            template.convertAndSend("/topic/message/" + updatedSession.name, it)
        }
        return sessionConverter.convertFromSession(updatedSession)
    }

    fun deleteSession(id: UUID) {
        val foundSession = sessionRepository.findById(id).getOrNull() ?: throw SessionNotFoundException(id.toString())
        logger.info("Deleting session, id: ${foundSession.id}, name: ${foundSession.name}")
        sessionRepository.deleteById(id)
    }

    companion object {
        val logger: Logger = Logger.getLogger(this::class.java.name)
    }
}
